AGENTS.md

```diff
@@ -0,0 +1,375 @@
+
+- Configure in `src/features.ts`
+- Access via `http://localhost:3000/#/frontend_features`
+- Manage locally: `localStorage.setItem('flags', "")`
+
+### Common development tasks
+
+- **Add new API**: Update swagger specs, run `npm run apis`
+- **Update proto definitions**: Modify protos, run respective build commands
+- **Add new component**: Create in `atoms/` or `components/`, add tests and stories
+- **Debug server**: Use `npm run start:proxy-and-server-inspect`
+- **Bundle analysis**: `npm run analyze-bundle`
+
+### Troubleshooting
+
+- **Port conflicts**: Frontend uses 3000 (React), 3001 (Node server), 3002 (API proxy)
+- **Node version issues**: Ensure you're using the version in `.nvmrc`
+- **API generation failures**: Check that swagger-codegen-cli.jar is in PATH
+- **Proto generation**: Requires `protoc` and `protoc-gen-grpc-web` in PATH
+- **Mock backend**: Limited API support; use real cluster for full testing
+
+## Lint and formatting checks
+
+- Go lint (CI uses `golangci-lint`):
+
+```bash
+golangci-lint run
+```
+
+- Python SDK import/order and unused import cleanups:
+
+```bash
+pip install -r sdk/python/requirements-dev.txt
+pycln --check sdk/python
+isort --check --profile google sdk/python
+```
+
+- Python SDK formatting (YAPF + string fixer):
+
+```bash
+pip install yapf pre_commit_hooks
+python3 -m pre_commit_hooks.string_fixer $(find sdk/python/kfp/**/*.py -type f)
+yapf --recursive --diff sdk/python/
+```
+
+- Python SDK docstring formatting:
+
+```bash
+pip install docformatter
+docformatter --check --recursive sdk/python/ --exclude "compiler_test.py"
+```
+
+## Common agent workflows
+
+- **Modify pipeline spec schema**:
+  1. Edit Protobufs under `api/`
+  2. Regenerate: `make -C api python && make -C api golang`
+  3. Update SDK/backend usages as needed
+- **Adjust Kubernetes behavior for tasks**:
+  - Resource requests/limits: set on component specs; the Driver converts these into pod spec patches.
+  - All other Kubernetes config: handled via `kubernetes_platform` platform spec.
+
+## Quick reference
+
+### Essential commands
+
+- Compile pipeline: `kfp dsl compile --py pipeline.py --output pipeline.yaml`
+- Generate protos: `make -C api python && make -C api golang`
+- Run SDK tests: `pytest -v sdk/python/kfp`
+- Run backend tests: `go test -v ./backend/...`
+- Check formatting:
+  `yapf --recursive --diff sdk/python/ && pycln --check sdk/python && isort --check --profile google sdk/python`
+- Frontend dev server: `cd frontend && npm start`
+- Frontend with cluster: `cd frontend && npm run start:proxy-and-server`
+- Frontend tests: `cd frontend && npm test`
+- Frontend formatting: `cd frontend && npm run format`
+- Generate frontend APIs: `cd frontend && npm run apis`
+
+### Key environment variables
+
+- `_KFP_RUNTIME=true`: Disables SDK imports during task execution
+- `LOCAL_API_SERVER=true`: Enables local API server testing mode when running integration tests on a Kind cluster
+
```

Comment:
We could add a section on clear indicators for auto-generated files. AI agents may try to edit generated files directly. `api/v2alpha1/python/kfp/pipeline_spec/pipeline_spec_pb2.py` for instance, is generated via a .proto, `api/v2alpha1/pipeline_spec.proto`.

Clearly specifying a `### 🚫 NEVER EDIT DIRECTLY (Generated Files)` section for instance, and adding such files in would be helpful. We could mention both the source files, and resulting auto-generated files here.